jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java

```diff
@@ -85,44 +86,43 @@ DbMetrics getDbMetrics() {
       if (metricsTrackingEnabled) {
         queryMetrics.update(executionTime, Thread.currentThread().getStackTrace()[3]);
       }
+
+      // We don't return the connection in a finally block because we don't want to
+      // do so in the case that a Throwable is thrown. In that case, the JVM is almost certainly
+      // about to exit, so returning the connection isn't important and in some cases can actually
+      // cause harm like committing a half-complete transaction (part of returning the connection is setting
+      // its autocommit property to true, which would commit a half-complete transaction).
+      dbManager.returnConnection(connection);
       return value;
     } catch (Exception e) {
       LOG.error("SQL execution failure", e);
       if (asTransaction) {
         connection.rollback();
       }
-      throw new SqlExecutionFailureException(e);
-    } finally {
-      // connection should be reset in PooledObjectFactory
+
       dbManager.returnConnection(connection);
-    }
-  }
+      throw new SqlExecutionFailureException(e);
+    } catch (Throwable t) {
 
-  private void execute(IExecution<DB> execution, boolean asTransaction) {
-    DB connection = dbManager.getConnection();
-    try {
-      connection.setAutoCommit(!asTransaction);
-      long startTime = System.currentTimeMillis();
-      execution.execute(connection);
-      if (asTransaction) {
-        connection.commit();
-      }
-      long executionTime = System.currentTimeMillis() - startTime;
-      if (metricsTrackingEnabled) {
-        queryMetrics.update(executionTime, Thread.currentThread().getStackTrace()[3]);
-      }
-    } catch (Exception e) {
-      LOG.error("SQL execution failure", e);
+      // We still try to explicitly rollback the transaction if a throwable is thrown.
       if (asTransaction) {
         connection.rollback();
       }
-      throw new SqlExecutionFailureException(e);
-    } finally {
-      // connection should be reset in PooledObjectFactory
-      dbManager.returnConnection(connection);
+
+      throw t;
     }
   }
 
+  private void execute(IExecution<DB> execution, boolean asTransaction) {
+    query(
+        db -> {
+          execution.execute(db);
+          return Optional.empty();
```

Comment:
```suggestion
          return null;
```
Nitpick: returning `null` also works, and is simpler.